### DISCUSSION TITLE
Fail to mount an OSS folder with an underline folder name

### DISCUSSION THREAD

**Discussion by wankunde** [juicedata/juicefs#278] [General]
**What happened**:

Fail to mount an OSS folder with an underline folder name
![image](https://user-images.githubusercontent.com/3626747/112956206-cad22100-9172-11eb-92ba-edc897157d70.png)

**What you expected to happen**:

Can mount an OSS folder with an underline folder name

![image](https://user-images.githubusercontent.com/3626747/112955812-6616c680-9172-11eb-937f-f04438de728d.png)

**How to reproduce it (as minimally and precisely as possible)**:

Try to mount an OSS folder with an underline folder name

**Anything else we need to know?**

**Environment**:
- JuiceFS version (use `./juicefs --version`) or Hadoop Java SDK version:  juicefs version 0.11.0 (2021-03-01 a354b982)
- Cloud provider or hardware configuration running JuiceFS: ECS
- OS (e.g: `cat /etc/os-release`): CentOS Linux
- Kernel (e.g. `uname -a`):  3.10.0-957.1.3.el7.x86_64
- Object storage (cloud provider and region): OSS
- Redis info (version, cloud provider managed or self maintained): redis-6.0
- Network connectivity (JuiceFS to Redis, JuiceFS to object storage): local network
- Others:

![image](https://user-images.githubusercontent.com/3626747/112956686-4c29b380-9173-11eb-8cfa-d5d803573144.png)

**Comment by xiaogaozi** [comment]
One JuiceFS volume corresponds to one OSS bucket. You only need format a volume for one OSS bucket once. So in your example, the OSS bucket is `xxx-offline-backup`, and your JuiceFS volume is already formatted. There's no need format a new volume for `prd_data_logs` subdirectory.

By the way, the last parameter of [`juicefs format`](https://github.com/juicedata/juicefs/blob/main/docs/en/command_reference.md#juicefs-format) command is just the volume name. It can be any value, and no connection with the name of subdirectory in the bucket.

**Comment by wankunde** [comment]
Hi, @xiaogaozi 

The OSS data directory is  `oss://leyan-offline-backup/custom-name/` in  the test log ?
![image](https://user-images.githubusercontent.com/3626747/112962866-69fa1700-9179-11eb-854b-268e96c73a04.png)

**Comment by xiaogaozi** [comment]
Yes, JuiceFS will create a new directory with the same name as volume in the bucket. For example, you format a volume called `test` and the OSS bucket is `leyan-offline-backup`, then the data will be stored into `oss://leyan-offline-backup/test` directory.

According to the screenshot, it seems like your access key and secret key are incorrect, so you don't have permission to access the bucket.

> You only need format a volume for one OSS bucket once.

Sorry my description is not accurate enough, I need to correct this sentence. Actually, you could format multiple volumes for one OSS bucket, but must specify different Redis address and volume name.

**Comment by wankunde** [comment]
I see, thank you @xiaogaozi

**Comment by wankunde** [comment]
> Yes, JuiceFS will create a new directory with the same name as volume in the bucket. 

Hi, @xiaogaozi  could you tell me why the volume name can't contains underscores ?
Users can create a directory with an underline folder name.

**Comment by wankunde** [comment]
Hi @xiaogaozi , I have another question:

I have read-only access to an OSS directory. Can I use juicefs to speed up reading.

```
./juicefs format \
>   --storage oss \
>   --bucket https://[bucket].[endpoint] \
>   --access-key test_key \
>   --secret-key test_value \
>   redis://:password@192.168.2.1:6379/1 \
>   prd_data_logs
2021/03/30 21:34:45.237871 juicefs[4015] <INFO>: Meta address: redis://:password@192.168.2.1:26379/1
2021/03/30 21:34:45.241813 juicefs[4015] <INFO>: Data uses oss://[bucket]/prd_data_logs/
2021/03/30 21:34:57.342776 juicefs[4015] <FATAL>: Storage oss://[bucket]/prd_data_logs/ is not configured correctly: Failed to put: oss: service returned error: StatusCode=403, ErrorCode=AccessDenied, ErrorMessage="You have no right to access this object because of bucket acl.", RequestId=606328FA0205A43239C1F87A
```

**Comment by allwefantasy** [comment]
> Hi @xiaogaozi , I have another question:
> 
> I have read-only access to an OSS directory. Can I use juicefs to speed up reading.
> 
> ```
> ./juicefs format \
> >   --storage oss \
> >   --bucket https://[bucket].[endpoint] \
> >   --access-key test_key \
> >   --secret-key test_value \
> >   redis://:password@192.168.2.1:6379/1 \
> >   prd_data_logs
> 2021/03/30 21:34:45.237871 juicefs[4015] <INFO>: Meta address: redis://:password@192.168.2.1:26379/1
> 2021/03/30 21:34:45.241813 juicefs[4015] <INFO>: Data uses oss://[bucket]/prd_data_logs/
> 2021/03/30 21:34:57.342776 juicefs[4015] <FATAL>: Storage oss://[bucket]/prd_data_logs/ is not configured correctly: Failed to put: oss: service returned error: StatusCode=403, ErrorCode=AccessDenied, ErrorMessage="You have no right to access this object because of bucket acl.", RequestId=606328FA0205A43239C1F87A
> ```

It seems impossible for now to speed up data already exists in OSS. Since JuiceFS will take control the meta data and you 
 can only see some special directory/file only JuiceFS can understand, this means JuiceFS is not a proxy.
